### PR TITLE
Add bruteforce and support compiling on Mac M1.

### PR DESCRIPTION
**Pull request by soothing-rain**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,22 +27,45 @@ if (NOT CMAKE_INSTALL_PREFIX)
...
+# TODO: Distinguish Apple M1 chip from other Apple chips.
+IF (APPLE)
     set(MACOS TRUE)
-    set(CMAKE_CXX_COMPILER "/usr/local/opt/llvm/bin/clang++")
-    set(CMAKE_CXX_FLAGS "-I/usr/local/opt/llvm/include -I/usr/local/include -I/usr/local/opt/openblas/include -I/usr/local/opt/libomp/include")
-    check_cxx_compiler_flag(-std=c++11 HAS_STD_CPP11_FLAG)
-    if(HAS_STD_CPP11_FLAG)
-     add_compile_options(-std=c++11)
+    set(CMAKE_C_COMPILER "/opt/homebrew/Cellar/llvm/13.0.0_2/bin/clang")
+    set(CMAKE_CXX_COMPILER "/opt/homebrew/Cellar/llvm/13.0.0_2/bin/clang++")
```
Comment (jaime0815) [inline CMakeLists.txt:36]: Do not use a specific path that may not work on other laptops.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt [hunk 2]
```diff
@@ -1,13 +1,9 @@
 cmake_minimum_required(VERSION 3.10)
 
-set (CMAKE_CXX_STANDARD 11)
-
-project(BruteForce)
-
-#set(CMAKE_C_COMPILER "/opt/homebrew/Cellar/llvm/13.0.0_2/bin/clang")
```
Comment (jaime0815) [inline knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt:7]: Do not use a specific path that may not work on other laptops.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/distances/CMakeLists.txt [hunk 3]
```diff
@@ -1,3 +1,3 @@
-set (CMAKE_CXX_STANDARD 11)
+set (CMAKE_CXX_STANDARD 17)
```
Comment (jaime0815) [inline knowhere/knowhere/index/vector_index/impl/bruteforce/distances/CMakeLists.txt:1]: guarantee CXX version is compatible with multiple platforms.

File: knowhere/CMakeLists.txt [hunk 4]
```diff
@@ -14,6 +14,7 @@
 include_directories(${INDEX_SOURCE_DIR}/knowhere)
 include_directories(${INDEX_SOURCE_DIR}/thirdparty)
 add_subdirectory(${INDEX_SOURCE_DIR}/knowhere/knowhere/utils)
+add_subdirectory(${INDEX_SOURCE_DIR}/knowhere/knowhere/index/vector_index/impl/bruteforce)
```
Comment (jaime0815) [inline knowhere/CMakeLists.txt:17]: bruteforce already is sub dir of knowwhere,  so using add_subdirectory is a best practice?

File: knowhere/knowhere/index/vector_index/impl/bruteforce/include/CMakeLists.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
-set (CMAKE_CXX_STANDARD 11)
```
Comment (jaime0815) [inline knowhere/knowhere/index/vector_index/impl/bruteforce/include/CMakeLists.txt:1]: guarantee CXX version is compatible with multiple platforms.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/indexing/CMakeLists.txt [hunk 6]
```diff
@@ -1,3 +1,3 @@
-set (CMAKE_CXX_STANDARD 11)
+set (CMAKE_CXX_STANDARD 17)
```
Comment (jaime0815) [inline knowhere/knowhere/index/vector_index/impl/bruteforce/indexing/CMakeLists.txt:1]: guarantee CXX version is compatible with multiple platforms.

File: CMakeLists.txt [hunk 7]
```diff
@@ -27,22 +27,45 @@ if (NOT CMAKE_INSTALL_PREFIX)
     set(CMAKE_INSTALL_PREFIX "./cmake_build")
 endif (NOT CMAKE_INSTALL_PREFIX)
 
-if (${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
-    message(STATUS "==============OSX Environment================")
+OPTION (USE_OpenMP "Use OpenMP to enable <omp.h>" ON)
+
+# TODO: Distinguish Apple M1 chip from other Apple chips.
+IF (APPLE)
```
Comment (jaime0815) [inline CMakeLists.txt:33]: to refer CMakeLists of root dir,  use MACOS as a variable.

File: knowhere/knowhere/utils/CMakeLists.txt [hunk 8]
```diff
@@ -14,34 +14,43 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-set(UTILS_SRC
-        distances_simd.cpp
-        FaissHookFvec.cpp
-        )
+if (APPLE)
```
Comment (jaime0815) [inline knowhere/knowhere/utils/CMakeLists.txt:17]: to refer CMakeLists of root dir,  use MACOS as a variable.

File: knowhere/knowhere/utils/CMakeLists.txt [hunk 9]
```diff
@@ -14,34 +14,43 @@
...
-        ${UTILS_SRC}
-        $<TARGET_OBJECTS:utils_avx>
-        $<TARGET_OBJECTS:utils_avx512>
-        )
+    add_library(knowhere_utils STATIC
+            ${UTILS_SRC}
+            $<TARGET_OBJECTS:utils_avx>
+            $<TARGET_OBJECTS:utils_avx512>
+            )
+endif ()
```
Comment (jaime0815) [inline knowhere/knowhere/utils/CMakeLists.txt:53]: where is compiled code block of  knowhere_utils  for linux?

**Comment by jaime0815** [comment]
Make sure both Mac and Linux can run this patch.

File: knowhere/CMakeLists.txt (hunk 4 above)
Comment (jaime0815) [reply knowhere/CMakeLists.txt:17]: cc @cydrain  any ideas?

Comment (cydrain) [reply knowhere/CMakeLists.txt:17]: need not add_subdirectory here, you can reference how nsg does

File: knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt [hunk 10]
```diff
@@ -1,13 +1,9 @@
 cmake_minimum_required(VERSION 3.10)
```
Comment (cydrain) [inline knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt:1]: please refer to nsg, we need not add bruteforce as a sub-project, so we need not add CMakeLists.txt here.

File: knowhere/knowhere/utils/CMakeLists.txt (hunk 9 above)
Comment (soothing-rain) [reply knowhere/knowhere/utils/CMakeLists.txt:53]: Linux cases are covered in the else block

File: knowhere/knowhere/utils/CMakeLists.txt (hunk 8 above)
Comment (soothing-rain) [reply knowhere/knowhere/utils/CMakeLists.txt]: done.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/indexing/CMakeLists.txt (hunk 6 above)
Comment (soothing-rain) [reply knowhere/knowhere/index/vector_index/impl/bruteforce/indexing/CMakeLists.txt]: This comment is removed. However could not find a liunx machine to test it out.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/include/CMakeLists.txt (hunk 5 above)
Comment (soothing-rain) [reply knowhere/knowhere/index/vector_index/impl/bruteforce/include/CMakeLists.txt]: ditto.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/distances/CMakeLists.txt (hunk 3 above)
Comment (soothing-rain) [reply knowhere/knowhere/index/vector_index/impl/bruteforce/distances/CMakeLists.txt]: ditto.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt (hunk 2 above)
Comment (soothing-rain) [reply knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt]: done.

File: knowhere/CMakeLists.txt (hunk 4 above)
Comment (soothing-rain) [reply knowhere/CMakeLists.txt:17]: done.

File: knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt (hunk 10 above)
Comment (soothing-rain) [reply knowhere/knowhere/index/vector_index/impl/bruteforce/CMakeLists.txt:1]: I kept the CMakeLists.txt for bruteforce directory, so as to run demo tests.

File: CMakeLists.txt (hunk 1 above)
Comment (soothing-rain) [reply CMakeLists.txt]: done.

File: CMakeLists.txt (hunk 7 above)
Comment (soothing-rain) [reply CMakeLists.txt]: done.

File: CMakeLists.txt [hunk 11]
```diff
@@ -30,11 +30,11 @@ endif (NOT CMAKE_INSTALL_PREFIX)
 if (${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
     message(STATUS "==============OSX Environment================")
     set(MACOS TRUE)
-    set(CMAKE_CXX_COMPILER "/usr/local/opt/llvm/bin/clang++")
-    set(CMAKE_CXX_FLAGS "-I/usr/local/opt/llvm/include -I/usr/local/include -I/usr/local/opt/openblas/include -I/usr/local/opt/libomp/include")
+    # Adjust accordingly if libomp is not install here.
+    set(CMAKE_CXX_FLAGS "-I/opt/homebrew/Cellar/libomp/13.0.0/include")
```
Comment (jaime0815) [inline CMakeLists.txt:34]: how to ensure this path work on other mac laptops?

Comment (soothing-rain) [reply CMakeLists.txt]: Forgot to change back.

Fixed.

**Pull request merged**